File: frontend/src/services/researchService.ts
import type { ResearchRequest, StreamMessage, ContentBrief } from '../types';
import { config, getApiUrl } from '../config';

// Custom fetch wrapper to always include ngrok bypass headers
const apiFetch = async (input: RequestInfo | URL, init?: RequestInit) => {
  const customInit = { ...init };
  customInit.headers = {
    ...customInit.headers,
    'ngrok-skip-browser-warning': 'true',
  };
  return fetch(input, customInit);
};

async function streamFetch(
  url: string,
  request: object,
  onMessage: (msg: StreamMessage) => void,
  onError: (err: string) => void
): Promise<void> {
  try {
    console.log('API call:', url);
    const response = await apiFetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(request),
    });
    console.log('API response:', response);

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }

    if (!response.body) {
      throw new Error('No response body');
    }

    const reader = response.body.getReader();
    const decoder = new TextDecoder('utf-8');
    let buffer = '';

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      // Decode chunk and add to buffer
      buffer += decoder.decode(value, { stream: true });

      // Process complete lines
      const lines = buffer.split('\n');
      buffer = lines.pop() || ''; // Keep incomplete line in buffer

      for (const line of lines) {
        if (line.trim() === '') continue;

        try {
          const data = JSON.parse(line) as StreamMessage;
          onMessage(data);
        } catch (e) {
          console.error('Failed to parse JSON line:', line, e);
        }
      }
    }

    // Process any remaining buffer
    if (buffer.trim()) {
      try {
        const data = JSON.parse(buffer) as StreamMessage;
        onMessage(data);
      } catch (e) {
        console.error('Failed to parse final buffer:', buffer, e);
      }
    }
  } catch (err) {
    const errorMsg = err instanceof Error ? err.message : String(err);
    onError(errorMsg);
  }
}

export const researchService = {
  /**
   * Main endpoint — handles chat, knowledge, and initial research classification
   */
  async callStageAResearch(
    request: ResearchRequest,
    onMessage: (msg: StreamMessage) => void,
    onError: (err: string) => void
  ): Promise<void> {
    const url = getApiUrl(config.api.endpoints.stageAResearch);
    return streamFetch(url, request, onMessage, onError);
  },

  /**
   * Marketing form submission — runs full Stage A pipeline (skips intent classification)
   */
  async callMarketingResearch(
    request: ResearchRequest,
    onMessage: (msg: StreamMessage) => void,
    onError: (err: string) => void
  ): Promise<void> {
    const url = getApiUrl(config.api.endpoints.stageAResearch + '/marketing');
    return streamFetch(url, request, onMessage, onError);
  },

  /**
   * Stage B — Generate marketing strategy from Stage A report
   */
  async callStageBStrategy(
    request: {
      stage_a_report: Record<string, unknown>;
      stage_a_input: Record<string, unknown>;
      mongodb_id?: string;
    },
    onMessage: (msg: StreamMessage) => void,
    onError: (err: string) => void
  ): Promise<void> {
    const url = getApiUrl(config.api.endpoints.stageBStrategy);
    return streamFetch(url, request, onMessage, onError);
  },

  /**
   * Stage B — Approve strategy and briefs
   */
  async approveStageBBriefs(
    request: {
      mongodb_id?: string;
      strategy: Record<string, unknown>;
      approved_briefs: ContentBrief[];
    }
  ): Promise<{ status: string; message: string; strategy_id?: string }> {
    const url = getApiUrl(config.api.endpoints.stageBApprove);
    try {
      const response = await fetch(url, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(request),
      });
      return await response.json();
    } catch (err) {
      return { status: 'error', message: String(err) };
    }
  },

  /**
   * Stage C — Execute campaign (image gen + Discord posting)
   */
  async callStageCCampaign(
    request: {
      approved_briefs: ContentBrief[];
      webhook_url?: string;
      image_api_url?: string;
      skip_image_generation?: boolean;
      mongodb_stage_a_id?: string;
    },
    onMessage: (msg: StreamMessage) => void,
    onError: (err: string) => void
  ): Promise<void> {
    const url = getApiUrl(config.api.endpoints.stageCCampaign);
    return streamFetch(url, request as object, onMessage, onError);
  },

  /**
   * Stage C — Schedule campaign (with specific posting times)
   */
  async callStageCCampaignScheduled(
    request: {
      approved_briefs: ContentBrief[];
      scheduled_times: string[]; // ISO 8601 datetimes
      webhook_url?: string;
      image_api_url?: string;
      skip_image_generation?: boolean;
      mongodb_stage_a_id?: string;
    },
    onMessage: (msg: StreamMessage) => void,
    onError: (err: string) => void
  ): Promise<void> {
    const url = getApiUrl(config.api.endpoints.stageCCampaign + '/scheduled');
    return streamFetch(url, request as object, onMessage, onError);
  },

  /**
   * Get scheduler status
   */
  async getSchedulerStatus(): Promise<{
    running: boolean;
    pending_briefs: number;
    check_interval: number;
  } | null> {
    try {
      const url = getApiUrl('/api/stage-c/scheduler/status');
      const response = await apiFetch(url);
      if (response.ok) {
        const data = await response.json();
        return data.data;
      }
      return null;
    } catch (err) {
      console.error('Failed to get scheduler status:', err);
      return null;
    }
  },

  /**
   * Get all scheduled campaigns
   */
  async getScheduledCampaigns(status?: string): Promise<any[]> {
    try {
      const url = status
        ? getApiUrl(`/api/stage-c/scheduler/campaigns?status=${status}`)
        : getApiUrl('/api/stage-c/scheduler/campaigns');
      const response = await apiFetch(url);
      if (response.ok) {
        const data = await response.json();
        return data.data?.campaigns || [];
      }
      return [];
    } catch (err) {
      console.error('Failed to get scheduled campaigns:', err);
      return [];
    }
  },

  /**
   * Get specific campaign details
   */
  async getCampaignDetails(campaignId: string): Promise<any | null> {
    try {
      const url = getApiUrl(`/api/stage-c/scheduler/campaigns/${campaignId}`);
      const response = await apiFetch(url);
      if (response.ok) {
        const data = await response.json();
        return data.data;
      }
      return null;
    } catch (err) {
      console.error('Failed to get campaign details:', err);
      return null;
    }
  },

  /**
   * Get pending briefs ready to post
   */
  async getPendingBriefs(): Promise<any> {
    try {
      const url = getApiUrl('/api/stage-c/scheduler/pending');
      const response = await apiFetch(url);
      if (response.ok) {
        const data = await response.json();
        return data.data;
      }
      return null;
    } catch (err) {
      console.error('Failed to get pending briefs:', err);
      return null;
    }
  },

  // ────────────────────────────────────────────────────────────
  // Conversation History APIs
  // ────────────────────────────────────────────────────────────

  /**
   * List all conversations
   */
  async listConversations(skip = 0, limit = 10): Promise<any> {
    try {
      const url = getApiUrl(`/api/conversations?skip=${skip}&limit=${limit}`);
      const response = await apiFetch(url);
      if (response.ok) {
        const data = await response.json();
        return data.data || { conversations: [], total: 0 };
      }
      return { conversations: [], total: 0 };
    } catch (err) {
      console.error('Failed to list conversations:', err);
      return { conversations: [], total: 0 };
    }
  },

  /**
   * Get a specific conversation
   */
  async getConversation(conversationId: string): Promise<any | null> {
    try {
      const url = getApiUrl(`/api/conversations/${conversationId}`);
      const response = await apiFetch(url);
      if (response.ok) {
        const data = await response.json();
        return data.data;
      }
      return null;
    } catch (err) {
      console.error('Failed to get conversation:', err);
      return null;
    }
  },

  /**
   * Create a new conversation
   */
  async createConversation(title?: string): Promise<any | null> {
    try {
      const url = getApiUrl('/api/conversations');
      const response = await apiFetch(url, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ title }),
      });
      if (response.ok) {
        const data = await response.json();
        return data.data;
      }
      return null;
    } catch (err) {
      console.error('Failed to create conversation:', err);
      return null;
    }
  },

  /**
   * Save messages to conversation
   */
  async saveMessagesToConversation(conversationId: string, messages: any[]): Promise<boolean> {
    try {
      const url = getApiUrl(`/api/conversations/${conversationId}/messages`);
      const response = await apiFetch(url, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ messages }),
      });
      return response.ok;
    } catch (err) {
      console.error('Failed to save messages:', err);
      return false;
    }
  },

  /**
   * Update conversation title
   */
  async updateConversationTitle(conversationId: string, title: string): Promise<boolean> {
    try {
      const url = getApiUrl(`/api/conversations/${conversationId}/title`);
      const response = await apiFetch(url, {
        method: 'PUT',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ title }),
      });
      return response.ok;
    } catch (err) {
      console.error('Failed to update title:', err);
      return false;
    }
  },

  /**
   * Delete a conversation
   */
  async deleteConversation(conversationId: string): Promise<boolean> {
    try {
      const url = getApiUrl(`/api/conversations/${conversationId}`);
      const response = await apiFetch(url, { method: 'DELETE' });
      return response.ok;
    } catch (err) {
      console.error('Failed to delete conversation:', err);
      return false;
    }
  },
};

File: android/app/src/main/java/com/example/marketmindai/ResearchService.java
package com.example.marketmindai;

import android.util.Log;
import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.google.gson.reflect.TypeToken;
import com.example.marketmindai.config.ApiConfig;
import com.example.marketmindai.model.ChatMessage;
import com.example.marketmindai.model.Conversation;
import com.example.marketmindai.model.StreamMessage;

import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;

/**
 * Research service for handling API calls to backend.
 * Supports streaming responses (NDJSON format).
 * 
 * API endpoints match frontend config.ts:
 *   /api/research/stage_a
 *   /api/research/stage_a/marketing
 *   /api/strategy/stage_b
 *   /api/strategy/stage_b/approve
 *   /api/campaign/stage_c
 *   /api/campaign/stage_c/scheduled
 *   /api/conversations
 */
public class ResearchService {
    private static final String TAG = "ResearchService";
    private static final OkHttpClient httpClient = new OkHttpClient.Builder()
            .readTimeout(5, TimeUnit.MINUTES)
            .writeTimeout(5, TimeUnit.MINUTES)
            .connectTimeout(30, TimeUnit.SECONDS)
            .build();
    
    private static final Gson gson = new Gson();
    
    /**
     * Callback interface for streaming responses
     */
    public interface StreamCallback {
        void onMessage(StreamMessage msg);
        void onError(String error);
        void onComplete();
    }
    
    /**
     * Error callback interface
     */
    public interface ErrorCallback {
        void onError(String error);
    }
    
    // ────────────────────────────────────────────────────────────
    // Streaming Research Endpoints
    // ────────────────────────────────────────────────────────────
    
    /**
     * Call Stage A research endpoint with streaming
     * Matches frontend: /api/research/stage_a
     */
    public static void callStageAResearch(
            android.content.Context context,
            Map<String, Object> request,
            StreamCallback callback,
            ErrorCallback errorCallback
    ) {
        String url = ApiConfig.getApiUrl("/api/research/stage_a");
        callStreamingEndpoint(context, url, request, callback, errorCallback);
    }
    
    /**
     * Call marketing research endpoint with streaming (skips intent classification)
     * Matches frontend: /api/research/stage_a/marketing
     */
    public static void callMarketingResearch(
            android.content.Context context,
            Map<String, Object> request,
            StreamCallback callback,
            ErrorCallback errorCallback
    ) {
        String url = ApiConfig.getApiUrl("/api/research/stage_a/marketing");
        callStreamingEndpoint(context, url, request, callback, errorCallback);
    }
    
    /**
     * Call Stage B strategy endpoint with streaming
     * Matches frontend: /api/strategy/stage_b
     */
    public static void callStageBStrategy(
            android.content.Context context,
            Map<String, Object> request,
            StreamCallback callback,
            ErrorCallback errorCallback
    ) {
        String url = ApiConfig.getApiUrl("/api/strategy/stage_b");
        callStreamingEndpoint(context, url, request, callback, errorCallback);
    }
    
    /**
     * Call Stage C campaign endpoint with streaming
     * Matches frontend: /api/campaign/stage_c
     */
    public static void callStageCCampaign(
            android.content.Context context,
            Map<String, Object> request,
            StreamCallback callback,
            ErrorCallback errorCallback
    ) {
        String url = ApiConfig.getApiUrl("/api/campaign/stage_c");
        callStreamingEndpoint(context, url, request, callback, errorCallback);
    }
    
    /**
     * Call Stage C scheduled campaign endpoint with streaming
     * Matches frontend: /api/campaign/stage_c/scheduled
     */
    public static void callStageCCampaignScheduled(
            android.content.Context context,
            Map<String, Object> request,
            StreamCallback callback,
            ErrorCallback errorCallback
    ) {
        String url = ApiConfig.getApiUrl("/api/campaign/stage_c/scheduled");
        callStreamingEndpoint(context, url, request, callback, errorCallback);
    }
    
    /**
     * Approve Stage B briefs (non-streaming)
     * Matches frontend: /api/strategy/stage_b/approve
     */
    public static void approveStageBBriefs(
            android.content.Context context,
            Map<String, Object> request
    ) {
        String url = ApiConfig.getApiUrl("/api/strategy/stage_b/approve");
        String authHeader = AuthService.getAuthHeader(context);
        
        Request.Builder builder = new Request.Builder()
                .url(url)
                .addHeader("ngrok-skip-browser-warning", "true")
                .post(RequestBody.create(
                        gson.toJson(request),
                        MediaType.parse("application/json")
                ));
        
        if (authHeader != null) {
            builder.addHeader("Authorization", authHeader);
        }
        
        try (Response response = httpClient.newCall(builder.build()).execute()) {
            if (!response.isSuccessful()) {
                Log.w(TAG, "Approve briefs failed: " + response.code());
            }
        } catch (IOException e) {
            Log.e(TAG, "Approve briefs error: " + e.getMessage());
        }
    }
    
    // ────────────────────────────────────────────────────────────
    // Conversation History APIs
    // Backend returns wrapped format: { "status": "success", "data": { ... } }
    // ────────────────────────────────────────────────────────────
    
    /**
     * List all conversations.
     * Response format: { "status": "...", "data": { "conversations": [...], "total": N } }
     */
    public static List<Conversation> listConversations(
            android.content.Context context,
            int skip,
            int limit
    ) {
        try {
            String url = ApiConfig.getApiUrl("/api/conversations?skip=" + skip + "&limit=" + limit);
            String authHeader = AuthService.getAuthHeader(context);
            Request.Builder builder = new Request.Builder()
                    .url(url)
                    .addHeader("ngrok-skip-browser-warning", "true")
                    .get();
            
            if (authHeader != null) {
                builder.addHeader("Authorization", authHeader);
            }
            
            Response response = httpClient.newCall(builder.build()).execute();
            String responseBody = response.body() != null ? response.body().string() : "";
            
            if (!response.isSuccessful()) {
                Log.w(TAG, "List conversations failed: " + response.code() + " - " + responseBody);
                return new ArrayList<>();
            }
            System.out.println("aaaaaaa jsonResponse");
            // Parse wrapped response: { "data": { "conversations": [...] } }
            JsonObject jsonResponse = gson.fromJson(responseBody, JsonObject.class);
            System.out.println(jsonResponse);
            if (jsonResponse != null && jsonResponse.has("data")) {
                JsonObject data = jsonResponse.getAsJsonObject("data");
                if (data != null && data.has("conversations")) {
                    JsonArray convArray = data.getAsJsonArray("conversations");
                    if (convArray != null) {
                        return gson.fromJson(convArray, new TypeToken<List<Conversation>>(){}.getType());
                    }
                }
            }
            
            return new ArrayList<>();
        } catch (Exception e) {
            Log.e(TAG, "List conversations error: " + e.getMessage());
            return new ArrayList<>();
        }
    }
    
    /**
     * Get a specific conversation with messages.
     * Response format: { "status": "...", "data": { "conversation_id": "...", "messages": [...] } }
     */
    public static Conversation getConversation(
            android.content.Context context,
            String conversationId
    ) {
        try {
            String url = ApiConfig.getApiUrl("/api/conversations/" + conversationId);
            String authHeader = AuthService.getAuthHeader(context);
            
            Request.Builder builder = new Request.Builder()
                    .url(url)
                    .addHeader("ngrok-skip-browser-warning", "true")
                    .get();
            
            if (authHeader != null) {
                builder.addHeader("Authorization", authHeader);
            }
            
            Response response = httpClient.newCall(builder.build()).execute();
            String responseBody = response.body() != null ? response.body().string() : "";
            
            if (!response.isSuccessful()) {
                Log.w(TAG, "Get conversation failed: " + response.code());
                return null;
            }
            
            // Parse wrapped response: { "data": { ... } }
            JsonObject jsonResponse = gson.fromJson(responseBody, JsonObject.class);
            if (jsonResponse != null && jsonResponse.has("data")) {
                JsonObject data = jsonResponse.getAsJsonObject("data");
                if (data != null) {
                    return gson.fromJson(data, Conversation.class);
                }
            }
            
            return null;
        } catch (Exception e) {
            Log.e(TAG, "Get conversation error: " + e.getMessage());
            return null;
        }
    }
    
    /**
     * Create a new conversation.
     * Response format: { "status": "...", "data": { "conversation_id": "..." } }
     */
    public static Conversation createConversation(
            android.content.Context context,
            String firstMessage,
            String title
    ) {
        try {
            String url = ApiConfig.getApiUrl("/api/conversations");
            String authHeader = AuthService.getAuthHeader(context);
            
            JsonObject requestBody = new JsonObject();
            if (firstMessage != null) requestBody.addProperty("first_message", firstMessage);
            if (title != null) requestBody.addProperty("title", title);
            
            Request.Builder builder = new Request.Builder()
                    .url(url)
                    .addHeader("ngrok-skip-browser-warning", "true")
                    .post(RequestBody.create(
                            gson.toJson(requestBody),
                            MediaType.parse("application/json")
                    ));
            
            if (authHeader != null) {
                builder.addHeader("Authorization", authHeader);
            }
            
            Response response = httpClient.newCall(builder.build()).execute();
            String responseBody = response.body() != null ? response.body().string() : "";
            
            if (!response.isSuccessful()) {
                Log.w(TAG, "Create conversation failed: " + response.code());
                return null;
            }
            
            // Parse wrapped response: { "data": { ... } }
            JsonObject jsonResponse = gson.fromJson(responseBody, JsonObject.class);
            if (jsonResponse != null && jsonResponse.has("data")) {
                JsonObject data = jsonResponse.getAsJsonObject("data");
                if (data != null) {
                    return gson.fromJson(data, Conversation.class);
                }
            }
            
            return null;
        } catch (Exception e) {
            Log.e(TAG, "Create conversation error: " + e.getMessage());
            return null;
        }
    }
    
    /**
     * Save messages to conversation
     */
    public static boolean saveMessagesToConversation(
            android.content.Context context,
            String conversationId,
            List<ChatMessage> messages
    ) {
        try {
            String url = ApiConfig.getApiUrl("/api/conversations/" + conversationId + "/messages");
            String authHeader = AuthService.getAuthHeader(context);
            
            JsonObject requestBody = new JsonObject();
            requestBody.add("messages", gson.toJsonTree(messages));
            
            Request.Builder builder = new Request.Builder()
                    .url(url)
                    .addHeader("ngrok-skip-browser-warning", "true")
                    .post(RequestBody.create(
                            gson.toJson(requestBody),
                            MediaType.parse("application/json")
                    ));
            
            if (authHeader != null) {
                builder.addHeader("Authorization", authHeader);
            }
            
            try (Response response = httpClient.newCall(builder.build()).execute()) {
                if (!response.isSuccessful()) {
                    Log.w(TAG, "Save messages failed: " + response.code());
                    return false;
                }
                return true;
            }
        } catch (Exception e) {
            Log.e(TAG, "Save messages error: " + e.getMessage());
            return false;
        }
    }
    
    /**
     * Delete a conversation
     */
    public static boolean deleteConversation(
            android.content.Context context,
            String conversationId
    ) {
        try {
            String url = ApiConfig.getApiUrl("/api/conversations/" + conversationId);
            String authHeader = AuthService.getAuthHeader(context);
            
            Request.Builder builder = new Request.Builder()
                    .url(url)
                    .addHeader("ngrok-skip-browser-warning", "true")
                    .delete();
            
            if (authHeader != null) {
                builder.addHeader("Authorization", authHeader);
            }
            
            try (Response response = httpClient.newCall(builder.build()).execute()) {
                if (!response.isSuccessful()) {
                    Log.w(TAG, "Delete conversation failed: " + response.code());
                    return false;
                }
                return true;
            }
        } catch (Exception e) {
            Log.e(TAG, "Delete conversation error: " + e.getMessage());
            return false;
        }
    }
    
    /**
     * Get scheduled campaigns
     */
    public static JsonArray getScheduledCampaigns(
            android.content.Context context,
            String status
    ) {
        try {
            String url = status != null
                    ? ApiConfig.getApiUrl("/api/stage-c/scheduler/campaigns?status=" + status)
                    : ApiConfig.getApiUrl("/api/stage-c/scheduler/campaigns");
            String authHeader = AuthService.getAuthHeader(context);
            
            Request.Builder builder = new Request.Builder()
                    .url(url)
                    .addHeader("ngrok-skip-browser-warning", "true")
                    .get();
            
            if (authHeader != null) {
                builder.addHeader("Authorization", authHeader);
            }
            
            Response response = httpClient.newCall(builder.build()).execute();
            String responseBody = response.body() != null ? response.body().string() : "";
            
            if (response.isSuccessful()) {
                JsonObject jsonResponse = gson.fromJson(responseBody, JsonObject.class);
                if (jsonResponse != null && jsonResponse.has("data")) {
                    JsonObject data = jsonResponse.getAsJsonObject("data");
                    if (data != null && data.has("campaigns")) {
                        return data.getAsJsonArray("campaigns");
                    }
                }
            }
            return new JsonArray();
        } catch (Exception e) {
            Log.e(TAG, "Get campaigns error: " + e.getMessage());
            return new JsonArray();
        }
    }
    
    // ────────────────────────────────────────────────────────────
    // Streaming Engine
    // ────────────────────────────────────────────────────────────
    
    /**
     * Generic streaming endpoint handler with NDJSON parsing.
     * Runs synchronously — must be called from a background thread.
     */
    private static void callStreamingEndpoint(
            android.content.Context context,
            String url,
            Map<String, Object> request,
            StreamCallback callback,
            ErrorCallback errorCallback
    ) {
        try {
            String authHeader = AuthService.getAuthHeader(context);
            
            Request.Builder builder = new Request.Builder()
                    .url(url)
                    .addHeader("ngrok-skip-browser-warning", "true")
                    .addHeader("Content-Type", "application/json")
                    .post(RequestBody.create(
                            gson.toJson(request),
                            MediaType.parse("application/json")
                    ));
            
            if (authHeader != null) {
                builder.addHeader("Authorization", authHeader);
            }
            
            Log.d(TAG, "API call: " + url);
            Response response = httpClient.newCall(builder.build()).execute();
            
            if (!response.isSuccessful()) {
                String errorMsg = "Unknown error";
                try {
                    if (response.body() != null) {
                        errorMsg = response.body().string();
                    }
                } catch (Exception e) {
                    Log.w(TAG, "Failed to read error response body: " + e.getMessage());
                }
                Log.e(TAG, "Request failed: " + response.code() + " - " + errorMsg);
                if (errorCallback != null) errorCallback.onError("Request failed: " + errorMsg);
                callback.onError("Request failed: " + response.code());
                return;
            }
            
            if (response.body() == null) {
                callback.onError("Empty response body");
                return;
            }
            
            // Process NDJSON stream line by line
            try (BufferedReader reader = new BufferedReader(
                    new InputStreamReader(response.body().byteStream())
            )) {
                String line;
                while ((line = reader.readLine()) != null) {
                    if (line.trim().isEmpty()) continue;
                    
                    try {
                        StreamMessage msg = gson.fromJson(line, StreamMessage.class);
                        if (msg != null) {
                            callback.onMessage(msg);
                        }
                    } catch (Exception e) {
                        Log.e(TAG, "Failed to parse stream message: " + e.getMessage());
                    }
                }
                callback.onComplete();
            }
        } catch (IOException e) {
            Log.e(TAG, "Stream error: " + e.getMessage());
            if (errorCallback != null) errorCallback.onError("Stream error: " + e.getMessage());
            callback.onError("Stream error: " + e.getMessage());
        }
    }
}
